Update license rules and license detections

AyanSinhaMahapatra opened the pull request:
### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁
* [x] Updated documentation pages (if applicable)
* [x] Updated CHANGELOG.rst (if applicable)
<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->

AyanSinhaMahapatra committed:
Add license rule for linux-syscall note

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on src/licensedcode/data/rules/false-positive_775.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: seen at kernel/bpf/syscall.c
+---
+
+sizeof(license) - 1) < 0)
```
This may be enough and also deals with the next one below?
```suggestion
sizeof(license)
```

pombredanne commented on src/licensedcode/data/rules/gpl-2.0-plus_or_bsd-new_22.RULE:
```diff
@@ -0,0 +1,9 @@
+---
+license_expression: gpl-2.0-plus OR bsd-new
+is_license_tag: yes
+relevance: 100
```
make it 99 because of the plain "BSD"

pombredanne commented on src/licensedcode/data/rules/false-positive_772.RULE:
```diff
@@ -0,0 +1,14 @@
+notes: seen in kernel /Documentation/process/license-rules.rst
+---
+
+The SPDX license identifier is added in form of a comment.  The comment
```
What about having multiple smaller, less specific FP rules for all the related FPs in this PR, here and below?
- SPDX license identifier is added
- SPDX-License-Identifier: <SPDX License
- (SPDX-License-Identifier) nor in
- SPDX license identifier in

pombredanne commented on src/licensedcode/data/rules/gcc-linking-exception-2.0_13.RULE:
```diff
@@ -0,0 +1,9 @@
+notes: seen in kernel /Documentation/process/license-rules.rst
+---
+
+SPDX-License-Identifier: <SPDX-License> WITH GCC-exception-2.0
```
See other suggestion above, make this an FP

pombredanne commented on src/licensedcode/data/rules/gpl-2.0_and_gpl-1.0-plus_and_gpl-2.0-plus_and_lgpl-2.0_and_lgpl-2.0-plus_and_lgpl-2.1_and_lgpl-2.1-plus_1.RULE:
```diff
@@ -0,0 +1,17 @@
+minimum_coverage: 90
+---
+
+The license described in the COPYING file applies to the kernel source
```
What about making this an FP?

AyanSinhaMahapatra replied:
This could ignore detections elsewhere then. it is best to detect correctly and ignore these as file category is different from code.

pombredanne commented on src/licensedcode/data/rules/license-clue_free-unknown_6.RULE:
```diff
@@ -0,0 +1,16 @@
+notes: seen in kernel /Documentation/process/license-rules.rst
+---
+
+Dual Licensing Only
```
Make this an FP

pombredanne commented on src/licensedcode/data/rules/license-clue_free-unknown_5.RULE:
```diff
@@ -0,0 +1,10 @@
+notes: seen in kernel /Documentation/process/license-rules.rst
+---
+
+Aside from that, individual files can be provided under a dual license,
```
Make this an FP

pombredanne commented on src/licensedcode/data/rules/linux-syscall-exception-gpl_7.RULE:
```diff
@@ -0,0 +1,14 @@
+notes: seen in linux kernel /LICENSES/exceptions/Linux-syscall-note
+---
+
+SPDX-Licenses: GPL-2.0, GPL-2.0+, GPL-1.0+, LGPL-2.0, LGPL-2.0+, LGPL-2.1, LGPL-2.1+, GPL-2.0-only, GPL-2.0-or-later
```
Make this an FP

pombredanne commented on src/licensedcode/data/rules/linux-syscall-exception-gpl_11.RULE:
```diff
@@ -0,0 +1,9 @@
+notes: seen in kernel /Documentation/process/license-rules.rst
+---
+
+SPDX-License-Identifier: <SPDX-License> WITH Linux-syscall-note
```
Make this an FP

pombredanne commented on src/licensedcode/data/rules/false-positive_776.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: seen at kernel/bpf/syscall.c
+---
+
+license[sizeof(license) - 1] = 0;
```
Like taken care of by a smaller combined FP with above

pombredanne reviewed (changes requested):
Thanks...  IMHO you cam make the FPs smaller and fewers... And you should FP most if not all of the doc parts that are comments about licensing

AyanSinhaMahapatra committed:
Modify license clue/false positive heuristics

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add false positive and other misc rules and updates

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add more rules and license detection improvements

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add new license and rules

Added by @DennisClark

Reference: https://github.com/aboutcode-org/scancode-toolkit/issues/3908
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add more license rules and fix test failures

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne committed:
Prefer exact license match over SPDX

Add a new matcher_order attribute to LicenseMatch and use it for sorting
matches rather than the matcher string.
This was we can ensure that there is a proper precedence between
matchers when two matches are matching exactly the same text.

The new sort order for matcher is like that:
- 0: 1-hash
- 1: 2-aho
- 2: 1-spdx-id
- 3: 3-seq
- 4: 5-undetected
- 5: 5-aho-frag
- 6: 6-unknown

The outcome is that a hash or aho match for the same text at the same
position will take precedence of the SPDX id match, allowing to curate
and correct some incorrect license expressions if needed.

Reference: https://github.com/aboutcode-org/scancode-toolkit/issues/3912
Reported-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented on src/licensedcode/data/rules/false-positive_784.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: found in linux/net/core/filter.c
+---
+
+EXPORT_PER_CPU_SYMBOL_GPL
```
Technically this is a GPL clue.

pombredanne commented on src/licensedcode/data/rules/false-positive_785.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: found in linux/include/linux/mfd/rn5t618.h
+---
+
+#define RN5T618_GPLED_FUNC
```
```suggestion
GPLED_FUNC
```

pombredanne commented on src/licensedcode/data/rules/false-positive_786.RULE:
```diff
@@ -0,0 +1,7 @@
+notes: found in linux/Documentation/process/maintainer-pgp-guide.rst
+---
+
+the more general "`Protecting Code Integrity`_" guide published by the
```
What are the issues here?

AyanSinhaMahapatra replied:
This was getting detected by https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/data/rules/lgpl-2.1-plus_53.RULE
```
{
          "license_expression": "lgpl-2.1-plus",
          "license_expression_spdx": "LGPL-2.1-or-later",
          "matches": [
            {
              "license_expression": "lgpl-2.1-plus",
              "spdx_license_expression": "LGPL-2.1-or-later",
              "from_file": "text",
              "start_line": 2,
              "end_line": 3,
              "matcher": "3-seq",
              "score": 18.18,
              "matched_length": 6,
              "match_coverage": 18.18,
              "rule_relevance": 100,
              "rule_identifier": "lgpl-2.1-plus_53.RULE",
              "rule_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/lgpl-2.1-plus_53.RULE",
              "matched_text": "the more general \"`Protecting Code Integrity`_\" guide published by the\nLinux Foundation. Please read that document for more in-depth discussion",
              "matched_text_diagnostics": "general \"`[Protecting] [Code] [Integrity]`_\" [guide] published by the\n[Linux] Foundation. Please"
            }
          ],
          "identifier": "lgpl_2_1_plus-ea16252f-118f-5f30-c0ae-734c9b7a5fe5"
        }
```

pombredanne commented on src/licensedcode/data/rules/false-positive_787.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: haha
+---
+
+sizeof(license) - 1) < 0)
```
```suggestion
sizeof(license)
```
... but 775 is there already, so this rule is redundant

pombredanne commented on src/licensedcode/data/rules/false-positive_788.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: haha
+---
+
+license[sizeof(license) - 1] = 0;
```
```suggestion
license[sizeof(license)
```

AyanSinhaMahapatra replied:
Not sure why, but changing this to the shorter false positive rule is not removing this detection, but the full rules are removing them allright

pombredanne commented on src/licensedcode/data/rules/gpl-2.0_1436.RULE:
```diff
@@ -0,0 +1,12 @@
+is_license_notice: yes
+---
+
+* Released under the terms of the {{GNU GPL v2.0.}}
```
Why not just have only "Released under the terms of the {{GNU GPL v2.0.}}" as a rule?

AyanSinhaMahapatra replied:
We are having a low score here otherwise as GNU is present and considered as license text? Maybe we should make `_GNU_SOURCE` a false positive?

pombredanne commented on src/licensedcode/data/rules/gpl-2.0_1437.RULE:
```diff
@@ -0,0 +1,13 @@
+is_license_notice: yes
+---
+
+and used with permission.
```
Why this part?
```suggestion
and used with permission.
```

AyanSinhaMahapatra replied:
lower score otherwise

pombredanne commented on src/licensedcode/data/rules/gpl-2.0_and_gpl-1.0-plus_and_gpl-2.0-plus_and_lgpl-2.0_and_lgpl-2.0-plus_and_lgpl-2.1_and_lgpl-2.1-plus_1.RULE:
```diff
@@ -0,0 +1,17 @@
+minimum_coverage: 90
+---
+
+The license described in the COPYING file applies to the kernel source
```
Are you sure of the license? the COPYING file in the kernel source would be GPL-2.0-only with Linux-syscall-notes
```suggestion
The license described in the {{COPYING file applies to the kernel source}}
```

AyanSinhaMahapatra replied:
lower score otherwise.

pombredanne commented on src/licensedcode/data/rules/false-positive_783.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: found in linux/drivers/pinctrl/samsung/pinctrl-exynos-arm.c
+---
+
+EXYNOS_PIN_BANK_EINTG("gpl1")
```
```suggestion
EINTG("gpl1")
```

pombredanne commented on src/licensedcode/data/rules/false-positive_782.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: found in linux/drivers/pinctrl/samsung/pinctrl-exynos-arm.c
+---
+
+EXYNOS_PIN_BANK_EINTG("gpl2")
```
```suggestion
EINTG("gpl2")
```

pombredanne commented on src/licensedcode/data/rules/false-positive_781.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: found in linux/drivers/gpu/drm/amd/include/asic_reg/nbio/
+---
+
+#define PCIE_RX_AD__RX_RC_UR_SSPL_MSG
```
```suggestion
UR_SSPL_MSG
```

pombredanne commented on src/licensedcode/data/rules/false-positive_776.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: seen in kernel /Documentation/process/license-rules.rst
+---
+
+ SPDX-License-Identifier: <SPDX License Expression>
```
This is also covered by 777 below and is not needed

pombredanne commented on src/licensedcode/data/rules/false-positive_773.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: seen in kernel /Documentation/process/license-rules.rst
+---
+
+(SPDX-License-Identifier) nor in any way relevant for expressing or
```
```suggestion
(SPDX-License-Identifier) nor 
```

pombredanne commented on src/licensedcode/data/rules/false-positive_771.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: seen in kernel /Documentation/process/license-rules.rst
+---
+
+The SPDX license identifier in kernel files shall be added at the first
```
```suggestion
The SPDX license identifier in kernel files
```

pombredanne commented on src/licensedcode/data/rules/false-positive_772.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: seen in kernel /Documentation/process/license-rules.rst
+---
+
+The SPDX license identifier is added in form of a comment.
```
```suggestion
The SPDX license identifier is added
```

pombredanne commented on src/licensedcode/data/rules/false-positive_774.RULE:
```diff
@@ -0,0 +1,6 @@
+notes: seen at drivers/block/floppy.c
+---
+
+* modeled after the freeware MS-DOS program fdformat/88 V1.8 by
```
```suggestion
* modeled after the freeware MS-DOS program
```

pombredanne committed:
Do not trace by default.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne reviewed (approved):
Here are a few nits for your consideration. I often prefer a shorter false positive. Good to merge otherwise.

pombredanne committed:
Add new license detection rule

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

AyanSinhaMahapatra committed:
Update rules and rule attributes

Reference: https://github.com/aboutcode-org/scancode-toolkit/issues/3922
Reported-by: Martin Ba @bilbothebaggins
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Refine false positive and license clue rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented:
@AyanSinhaMahapatra Thanks!
merging
